src/H5Dio.c

```diff
@@ -137,363 +131,597 @@ H5D__read(H5D_t *dataset, hid_t mem_type_id, H5S_t *mem_space, H5S_t *file_space
-        (*io_info.layout_ops.io_init)(&io_info, &type_info, nelmts, file_space, mem_space, fm) < 0)
-        HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "can't initialize I/O info")
-    io_op_init = TRUE;
+        /* Sanity check that space is allocated, if there are elements */
+        if (dset_info[i].nelmts > 0)
+            HDassert((*dset_info[i].dset->shared->layout.ops->is_space_alloc)(
+                         &dset_info[i].dset->shared->layout.storage) ||
+                     (dset_info[i].dset->shared->layout.ops->is_data_cached &&
+                      (*dset_info[i].dset->shared->layout.ops->is_data_cached)(dset_info[i].dset->shared)) ||
+                     dset_info[i].dset->shared->dcpl_cache.efl.nused > 0 ||
+                     dset_info[i].dset->shared->layout.type == H5D_COMPACT);
+
+        /* Call storage method's I/O initialization routine */
+        if (dset_info[i].layout_ops.io_init &&
+            (dset_info[i].layout_ops.io_init)(&io_info, &(dset_info[i])) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "can't initialize I/O info")
+        io_op_init++;
+
+        /* Reset metadata tagging */
+        H5AC_tag(prev_tag, NULL);
+    } /* end of for loop */
+
+    HDassert(type_info_init == count);
+    HDassert(io_op_init == count);
+
+    /* Perform second phase of type info initialization */
+    if (H5D__typeinfo_init_phase2(&io_info) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "unable to set up type info (second phase)")
 
 #ifdef H5_HAVE_PARALLEL
     /* Adjust I/O info for any parallel I/O */
-    if (H5D__ioinfo_adjust(&io_info, dataset, file_space, mem_space, &type_info) < 0)
+    if (H5D__ioinfo_adjust(&io_info) < 0)
         HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "unable to adjust I/O info for parallel I/O")
 #endif /*H5_HAVE_PARALLEL*/
 
-    /* Invoke correct "high level" I/O routine */
-    if ((*io_info.io_ops.multi_read)(&io_info, &type_info, nelmts, file_space, mem_space, fm) < 0)
-        HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read data")
+    /* If multi dataset I/O callback is not provided, perform read IO via
+     * single-dset path with looping */
+    if (io_info.md_io_ops.multi_read_md) {
+        /* Create sel_pieces array if any pieces are selected */
+        if (io_info.piece_count > 0) {
+            HDassert(!io_info.sel_pieces);
+            HDassert(io_info.pieces_added == 0);
+
+            /* Allocate sel_pieces array */
+            if (NULL ==
+                (io_info.sel_pieces = H5MM_malloc(io_info.piece_count * sizeof(io_info.sel_pieces[0]))))
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate array of selected pieces")
+        }
+
+        /* MDIO-specific second phase initialization */
+        for (i = 0; i < count; i++)
+            if (dset_info[i].layout_ops.mdio_init &&
+                (dset_info[i].layout_ops.mdio_init)(&io_info, &(dset_info[i])))
```

Comment:
`< 0` ? Though it'll still work fine as is now

Reply:
Fixed, thanks